packages/firestore/src/util/async_queue.ts

```diff
@@ -20,30 +20,103 @@ import { AnyDuringMigration, AnyJs } from './misc';
 import { Deferred } from './promise';
 import { Code, FirestoreError } from './error';
 
-type DelayedOperation<T> = {
+/** External Result of scheduling a delayed operation. */
+export interface DelayedOperationResult<T> {
+  /** A promise that will resolve once the operation has been run. */
+  promise: Promise<T>;
+
+  /**
+   * Prevents the operation from running and rejects the promise with a
+   * Code.CANCELLED error.
+   */
+  cancel(): void;
+}
+
+/**
+ * Represents an operation scheduled to be run in the future.
+ *
+ * Created via DelayedOperation.createAndSchedule().
+ * Supports cancellation (via cancel()) and early execution (via scheduleNow()).
+ */
+class DelayedOperation<T> implements DelayedOperationResult<T> {
   // tslint:disable-next-line:no-any Accept any return type from setTimeout().
-  handle: any;
-  op: () => Promise<T>;
-  deferred: Deferred<T>;
-};
+  private timerHandle: any;
+  private readonly deferred = new Deferred<T>();
+  /** true if the operation has not been executed or cancelled yet. */
+  private pending = true;
+
+  static createAndSchedule<T>(
+    asyncQueue: AsyncQueue,
+    op: () => Promise<T>,
+    delayMs: number
+  ) {
+    const delayedOp = new DelayedOperation(op);
+    delayedOp.timerHandle = setTimeout(
+      () => delayedOp.scheduleNow(asyncQueue),
+      delayMs
+    );
+    return delayedOp;
+  }
+
+  private constructor(private op: () => Promise<T>) {}
+
+  get promise(): Promise<T> {
+    return this.deferred.promise;
+  }
+
+  /**
+   * Schedules the operation to run on the provided AsyncQueue if it has not
+   * already been run or cancelled.
+   */
+  scheduleNow(asyncQueue: AsyncQueue): void {
+    this.clearTimeout();
+    asyncQueue.schedule(this.runIfNecessary.bind(this));
+  }
+
+  cancel(reason?: string): void {
+    if (this.pending) {
+      this.pending = false;
+      this.clearTimeout();
+      this.deferred.reject(
+        new FirestoreError(
+          Code.CANCELLED,
+          'Operation cancelled' + (reason ? ': ' + reason : '')
+        )
+      );
+    }
+  }
+
+  private runIfNecessary(): Promise<void> {
```

Comment:
Suggest to use `runIfPending` to match the names in the rest of this class.

Reply:
Merged this method into scheduleNow (which got renamed to handleDelayElapesd)